EDDiscovery/Actions/ActionCommands/ActionSay.cs

```diff
@@ -131,33 +151,42 @@ public override bool ExecuteAction(ActionProgramRun ap)
                     if (rate == -999)
                         rate = ap.currentvars.GetInt(globalvarspeechrate, 0);
 
+                    string culture = vars.ContainsKey(culturename) ? vars[culturename] : (ap.currentvars.ContainsKey(globalvarspeechculture) ? ap.currentvars[globalvarspeechculture] : "Default");
+
+
                     Audio.SoundEffectSettings ses = new Audio.SoundEffectSettings(vars);        // use the rest of the vars to place effects
 
                     if (!ses.Any && !ses.OverrideNone && ap.currentvars.ContainsKey(globalvarspeecheffects))  // if can't see any, and override none if off, and we have a global, use that
                     {
                         vars = new ConditionVariables(ap.currentvars[globalvarspeecheffects], ConditionVariables.FromMode.MultiEntryComma);
                     }
 
-                    string phrase = ap.actioncontroller.DiscoveryForm.SpeechSynthesizer.ToPhrase(say, out errlist, ap.functions, ap.currentvars);
-
-                    if (errlist == null)
+                    string expsay;
+                    if (ap.functions.ExpandString(say, ap.currentvars, out expsay) != EDDiscovery.ConditionLists.ExpandResult.Failed)
                     {
-                        if (phrase.Length == 0) // just abort..
-                            return true;
+                        expsay = expsay.PickOneOfGroups(rnd);
 
 #if true
-                        System.IO.MemoryStream ms = ap.actioncontroller.DiscoveryForm.SpeechSynthesizer.Speak(phrase, voice, rate);
+                        System.IO.MemoryStream ms = ap.actioncontroller.DiscoveryForm.SpeechSynthesizer.Speak(expsay, culture, voice, rate);
```

Comment:
I can't tell disposition looking at this on my phone, but memory streams are another prime candidate for `using()` (or explicit `Dispose()` when complete).